### PR TITLE
terror: provide a RegisterFinish API

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Prevent regression like https://github.com/pingcap/tidb/issues/28190

### What is changed and how it works?

TiDB will register all the error kinds in its initialization, and then call this API
After that, misuse of the terror will panic, thus the bug is exposed early.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

In TiDB

Code changes

 - Has exported function/method change

### CODE REVIEW

**Comment by tiancaiamao** [comment]
https://github.com/pingcap/tidb/pull/28438

File: terror/terror.go [hunk 1]
```diff
@@ -107,6 +108,18 @@ func (m *code2ErrClassMap) Put(key string, err ErrClass) {
 	m.data.Store(key, err)
 }
 
+var registerFinish uint32
+
+// RegisterFinish makes the register of new error panic.
+// The use pattern should be regist all the error during initialization, and then call RegisterFinish.
```
Comment (kennytm) [inline terror/terror.go:114]:
```suggestion
// The use pattern should be register all the errors during initialization, and then call RegisterFinish.
```

**Review by ichn-hu — APPROVED** [review]
LGTM

**Comment by tiancaiamao** [comment]
My main point is actually here https://github.com/pingcap/tidb/pull/25405#issuecomment-861173193

> Is there some way to prevent this kind of error in the future?
> Maybe nobody read the document and maybe people still make this kind of mistakes in the future.

We have got stuck on the error before, and fixed it then.
But later a contributor file a PR and misuse the terror package again...
This time we're not so lucky, the reviewers don't catch it and our customer trigger the bug

**Review by kennytm — APPROVED** [review]

**Pull request merged**